### PR TITLE
Use latest language syntax

### PR DESCRIPTION
**Pull request by AnselmC**
Use [pyugrade](https://github.com/asottile/pyupgrade) to make sure we use the latest language features for the target version. 
Not sure if this is controversial (aka seen as code golf) but IMO it enforces consistent style and is generally less verbose.

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/codegen/linearizer.py             1               1
tinygrad/graph.py                          2               2
------------------------------------------------------------
total                                      3               3
------------------------------------------------------------
lines added in the tinygrad folder: 0
```

File: docs/abstractions.py [hunk 1]
```diff
@@ -85,18 +85,18 @@ class LazyBuffer:
...
-  op: Optional[LazyOp]
+  op: LazyOp | None
 
 # LazyOp (in tinygrad/ops.py, code 4/10)
 # in a tree they form an Abstract Syntax Tree for a single GPU kernel
 class LazyOp:
   op: Op                                       # the type of the compute
-  src: Tuple[Union[LazyOp, LazyBuffer], ...]   # the sources
-  arg: Optional[Any] = None                    # and an optional static argument
+  src: tuple[LazyOp | LazyBuffer, ...]   # the sources
```
Comment (jla524) [inline docs/abstractions.py:98]: does this work on python 3.8?

**Comment by geohot** [comment]
tinygrad supports back to Python 3.8 for Ubuntu 20.04

**Comment by AnselmC** [comment]
> tinygrad supports back to Python 3.8 for Ubuntu 20.04

? all the updates in this PR are python 3.8 syntax

**Comment by geohot** [comment]
Err, so there's a chance I'm wrong about this.

I thought the `|` instead of `Union` was later.

**Comment by jla524** [comment]
> Err, so there's a chance I'm wrong about this. I thought the `|` instead of `Union` was later.

It was [introduced in python 3.10](https://peps.python.org/pep-0604/), but CI passed with the 3.8 runners thanks to `from __future__ import annotations`.

**Comment by AnselmC** [comment]
I totally agree that some of these are more code golf, however, I don't think you can pick-and-choose what `pyupgrade` does. 

I also agree that something like this needs to be enforced, hence the new `pre-commit` and gh workflow jobs.

In either case, lmk if you want to go ahead or not. If you do, I'll resolve the conflicts. If not, also good.

**Comment by jdlib** [comment]
> A bunch of these changes are also very code golf. Why are the rs being removed from open? I know it's the same but why change it?

depends on the definition of code golf. Why explicitly pass the default value?

**Comment by cauliyang** [comment]
If introducing `pyupdate`, why not use [ruff](https://github.com/astral-sh/ruff) directly?

**Comment by geohot** [comment]
Yea, going to close this, I shouldn't have nitpicked about python 3.8. From discord: "every repo get shoved into the style wars without constant vigilance"

How does this PR further the goals of the tinygrad project?

**Pull request closed without merging**